smcontroller/smcontroller.go

```diff
@@ -340,6 +316,89 @@ func (controller *Controller) RegisterSM(stream pb.SMService_RegisterSMServer) e
  * Private
  **********************************************************************************************************************/
 
+func (controller *Controller) getSystemLog(logRequest cloudprotocol.RequestLog) error {
+	if logRequest.Filter.NodeIDs != nil {
+		for _, nodeID := range logRequest.Filter.NodeIDs {
+			handler, err := controller.getNodeHandlerByID(nodeID)
+			if err != nil {
+				return err
+			}
+
+			if err := handler.getSystemLog(logRequest); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	for _, handler := range controller.nodes {
+		if err := handler.getSystemLog(logRequest); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (controller *Controller) getServiceLog(logRequest cloudprotocol.RequestLog) error {
+	if logRequest.Filter.ServiceID == nil {
+		return aoserrors.New("serviceId required field for service log")
+	}
+
+	if logRequest.Filter.NodeIDs != nil {
+		for _, nodeID := range logRequest.Filter.NodeIDs {
+			handler, err := controller.getNodeHandlerByID(nodeID)
+			if err != nil {
+				return err
+			}
+
+			if err := handler.getInstanceLog(logRequest); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	for _, handler := range controller.nodes {
+		if err := handler.getInstanceLog(logRequest); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (controller *Controller) getCrashLog(logRequest cloudprotocol.RequestLog) error {
+	if logRequest.Filter.ServiceID == nil {
+		return aoserrors.New("serviceId required field for service log")
+	}
+
+	if logRequest.Filter.NodeIDs != nil {
```

Comment:
logRequest.Filter.NodeIDs is arra, not a pointer. We have to check len

Reply:
`+`

Reply:
Done